File: src/main/java/org/hwabeag/hwaskyblock/inventorys/HwaSkyBlockSharerGUI.java
package org.hwabeag.hwaskyblock.inventorys;

import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.World;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.entity.Player;
import org.bukkit.event.Listener;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.inventory.meta.SkullMeta;
import org.hwabeag.hwaskyblock.config.ConfigManager;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class HwaSkyBlockSharerGUI implements Listener {
    private final Inventory inv;

    FileConfiguration Config = ConfigManager.getConfig("setting");
    FileConfiguration SkyBlockConfig = ConfigManager.getConfig("skyblock");
    FileConfiguration PlayerConfig = ConfigManager.getConfig("player");

    public HwaSkyBlockSharerGUI(Player player, String key) {
        inv = Bukkit.createInventory(null, 54, Objects.requireNonNull(Config.getString("gui-name.sharer_setting")));
        initItemSetting(player, key);
    }

    private ItemStack getHead(Player player, String name) {
        World world = player.getWorld();
        String world_name = world.getWorldFolder().getName();
        String[] number = world_name.split("\\.");
        String id = number[1];

        boolean player_join = SkyBlockConfig.getBoolean(id + ".sharer." + name + ".join");
        boolean block_break = SkyBlockConfig.getBoolean(id + ".sharer." + name + ".break");
        boolean block_place = SkyBlockConfig.getBoolean(id + ".sharer." + name + ".place");

        ItemStack item = new ItemStack(Material.PLAYER_HEAD, 1, (short) 3);
        SkullMeta skull = (SkullMeta) item.getItemMeta();
        @Nullable String display_name = Config.getString("gui-slot-item-name.sharer_setting.sharer");
        display_name = Objects.requireNonNull(display_name).replace("{name}", name);
        skull.setDisplayName(ChatColor.translateAlternateColorCodes('&', display_name));
        skull.setOwner(name);
        ArrayList<String> loreList = new ArrayList<>();
        Player player_exact = Bukkit.getServer().getPlayerExact(name);
        if (player_exact != null) {
            loreList.add(ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("gui-slot-item-name.sharer_setting.online-lore"))));
        } else {
            loreList.add(ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("gui-slot-item-name.sharer_setting.offline-lore"))));
        }

        loreList.add("");
        loreList.add(ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("gui-slot-item-name.sharer_setting.join"))));
        if (player_join) {
            loreList.add(ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("gui-slot-item-name.sharer_setting.join-true"))));
        } else {
            loreList.add(ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("gui-slot-item-name.sharer_setting.join-false"))));
        }

        loreList.add("");
        loreList.add(ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("gui-slot-item-name.sharer_setting.break"))));
        if (block_break) {
            loreList.add(ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("gui-slot-item-name.sharer_setting.break-true"))));
        } else {
            loreList.add(ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("gui-slot-item-name.sharer_setting.break-false"))));
        }

        loreList.add("");
        loreList.add(ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("gui-slot-item-name.sharer_setting.place"))));
        if (block_place) {
            loreList.add(ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("gui-slot-item-name.sharer_setting.place-true"))));
        } else {
            loreList.add(ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("gui-slot-item-name.sharer_setting.place-false"))));
        }

        loreList.add("");
        loreList.add(ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("gui-slot-item-name.sharer_setting.use"))));

        skull.setLore(loreList);
        item.setItemMeta(skull);
        return item;
    }

    private void initItemSetting(Player player, String id) {
        String name = player.getName();
        int N = 0;
        int Page = 1;
        if (SkyBlockConfig.getConfigurationSection(id + ".sharer") != null) {
            for (String key : Objects.requireNonNull(SkyBlockConfig.getConfigurationSection(id + ".sharer")).getKeys(false)) {
                int PlayerPage = PlayerConfig.getInt(name + ".skyblock.page");
                if (Page == PlayerPage) {
                    ItemStack item = getHead(player, key);
                    inv.setItem(N, item);
                }
                N = N + 1;
                if (N >= 44) {
                    Page = Page + 1;
                    N = 0;
                }
            }
        }

        ItemStack item = new ItemStack(Material.PAPER, 1);
        ItemMeta itemMeta = item.getItemMeta();
        itemMeta.setDisplayName(ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("gui-slot-item-name.previous_page"))));
        ArrayList<String> loreList = new ArrayList<>();
        for (String lore : Config.getStringList("gui-slot-item-name.previous_page-lore")) {
            loreList.add(ChatColor.translateAlternateColorCodes('&', lore));
        }
        itemMeta.setLore(loreList);
        item.setItemMeta(itemMeta);
        inv.setItem(45, item);


        item = new ItemStack(Material.PAPER, 1);
        itemMeta = item.getItemMeta();
        itemMeta.setDisplayName(ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("gui-slot-item-name.next_page"))));
        loreList = new ArrayList<>();
        for (String lore : Config.getStringList("gui-slot-item-name.next_page-lore")) {
            loreList.add(ChatColor.translateAlternateColorCodes('&', lore));
        }
        itemMeta.setLore(loreList);
        item.setItemMeta(itemMeta);
        inv.setItem(53, item);
    }

    public void open(Player player) {
        player.openInventory(inv);
    }

}

File: src/main/kotlin/inventorys/HwaSkyBlockSharerGUI.kt
package org.hwabaeg.hwaskyblock.inventorys

import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.Material
import org.bukkit.configuration.file.FileConfiguration
import org.bukkit.entity.Player
import org.bukkit.event.Listener
import org.bukkit.inventory.Inventory
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.SkullMeta
import org.hwabaeg.hwaskyblock.database.DatabaseManager
import org.hwabaeg.hwaskyblock.database.config.ConfigManager
import java.util.*

class HwaSkyBlockSharerGUI(player: Player, key: String?) : Listener {
    private val inv: Inventory

    var MessageConfig: FileConfiguration = ConfigManager.getConfig("message")!!

    init {
        inv = Bukkit.createInventory(
            null,
            54,
            Objects.requireNonNull<String?>(MessageConfig.getString("gui-name.sharer_setting"))
        )
        initItemSetting(player, key)
    }

    private fun getHead(player: Player, name: String): ItemStack {
        val world = player.world
        val world_name = world.worldFolder.getName()
        val number: Array<String?> = world_name.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val id = number[1]

        val player_join =
            DatabaseManager.getShareData("$id", name, "can_join", null) as? Boolean ?: false
        val block_break =
            DatabaseManager.getShareData("$id", name, "can_break", null) as? Boolean ?: false
        val block_place =
            DatabaseManager.getShareData("$id", name, "can_place", null) as? Boolean ?: false

        val item = ItemStack(Material.PLAYER_HEAD, 1, 3.toShort())
        val skull = item.itemMeta as SkullMeta
        var display_name = MessageConfig.getString("gui-slot-item-name.sharer_setting.sharer")
        display_name = Objects.requireNonNull<String?>(display_name).replace("{name}", name)
        skull.setDisplayName(ChatColor.translateAlternateColorCodes('&', display_name))
        skull.owner = name
        val loreList = ArrayList<String?>()
        val player_exact = Bukkit.getServer().getPlayerExact(name)
        if (player_exact != null) {
            loreList.add(
                ChatColor.translateAlternateColorCodes(
                    '&',
                    Objects.requireNonNull<String?>(MessageConfig.getString("gui-slot-item-name.sharer_setting.online-lore"))
                )
            )
        } else {
            loreList.add(
                ChatColor.translateAlternateColorCodes(
                    '&',
                    Objects.requireNonNull<String?>(MessageConfig.getString("gui-slot-item-name.sharer_setting.offline-lore"))
                )
            )
        }

        loreList.add("")
        loreList.add(
            ChatColor.translateAlternateColorCodes(
                '&',
                Objects.requireNonNull<String?>(MessageConfig.getString("gui-slot-item-name.sharer_setting.join"))
            )
        )
        if (player_join) {
            loreList.add(
                ChatColor.translateAlternateColorCodes(
                    '&',
                    Objects.requireNonNull<String?>(MessageConfig.getString("gui-slot-item-name.sharer_setting.join-true"))
                )
            )
        } else {
            loreList.add(
                ChatColor.translateAlternateColorCodes(
                    '&',
                    Objects.requireNonNull<String?>(MessageConfig.getString("gui-slot-item-name.sharer_setting.join-false"))
                )
            )
        }

        loreList.add("")
        loreList.add(
            ChatColor.translateAlternateColorCodes(
                '&',
                Objects.requireNonNull<String?>(MessageConfig.getString("gui-slot-item-name.sharer_setting.break"))
            )
        )
        if (block_break) {
            loreList.add(
                ChatColor.translateAlternateColorCodes(
                    '&',
                    Objects.requireNonNull<String?>(MessageConfig.getString("gui-slot-item-name.sharer_setting.break-true"))
                )
            )
        } else {
            loreList.add(
                ChatColor.translateAlternateColorCodes(
                    '&',
                    Objects.requireNonNull<String?>(MessageConfig.getString("gui-slot-item-name.sharer_setting.break-false"))
                )
            )
        }

        loreList.add("")
        loreList.add(
            ChatColor.translateAlternateColorCodes(
                '&',
                Objects.requireNonNull<String?>(MessageConfig.getString("gui-slot-item-name.sharer_setting.place"))
            )
        )
        if (block_place) {
            loreList.add(
                ChatColor.translateAlternateColorCodes(
                    '&',
                    Objects.requireNonNull<String?>(MessageConfig.getString("gui-slot-item-name.sharer_setting.place-true"))
                )
            )
        } else {
            loreList.add(
                ChatColor.translateAlternateColorCodes(
                    '&',
                    Objects.requireNonNull<String?>(MessageConfig.getString("gui-slot-item-name.sharer_setting.place-false"))
                )
            )
        }

        loreList.add("")
        loreList.add(
            ChatColor.translateAlternateColorCodes(
                '&',
                Objects.requireNonNull<String?>(MessageConfig.getString("gui-slot-item-name.sharer_setting.use"))
            )
        )

        skull.lore = loreList
        item.itemMeta = skull
        return item
    }

    private fun initItemSetting(player: Player, id: String?) {
        val name = player.name
        var N = 0
        var Page = 1
        val sharerData = DatabaseManager.getSharePermissionsMap("$id") as? Map<*, *>
        if (sharerData != null) {
            for (key in sharerData.keys) {
                val PlayerPage =
                    DatabaseManager.getUserData("$name.skyblock.page", player, "getSkyblockPage") as? Int ?: 0
                if (Page == PlayerPage) {
                    val item = getHead(player, key.toString())
                    inv.setItem(N, item)
                }
                N = N + 1
                if (N >= 44) {
                    Page = Page + 1
                    N = 0
                }
            }
        }

        var item = ItemStack(Material.PAPER, 1)
        var itemMeta = item.itemMeta
        itemMeta?.setDisplayName(
            ChatColor.translateAlternateColorCodes(
                '&',
                Objects.requireNonNull<String?>(MessageConfig.getString("gui-slot-item-name.previous_page"))
            )
        )
        var loreList = ArrayList<String?>()
        for (lore in MessageConfig.getStringList("gui-slot-item-name.previous_page-lore")) {
            loreList.add(ChatColor.translateAlternateColorCodes('&', lore))
        }
        itemMeta?.setLore(loreList)
        item.itemMeta = itemMeta
        inv.setItem(45, item)

        item = ItemStack(Material.PAPER, 1)
        itemMeta = item.itemMeta
        itemMeta?.setDisplayName(
            ChatColor.translateAlternateColorCodes(
                '&',
                Objects.requireNonNull<String?>(MessageConfig.getString("gui-slot-item-name.next_page"))
            )
        )
        loreList = ArrayList<String?>()
        for (lore in MessageConfig.getStringList("gui-slot-item-name.next_page-lore")) {
            loreList.add(ChatColor.translateAlternateColorCodes('&', lore))
        }
        itemMeta?.lore = loreList
        item.itemMeta = itemMeta
        inv.setItem(53, item)
    }

    fun open(player: Player) {
        player.openInventory(inv)
    }
}
